elastic/kibana
# Refactor use url state hook

## Summary

This PR aims to simplify `useUrlState` and improve readability by implementing quick and easy changes (low hanging fruits).

* There are no big refactor or any visual changes.
* It only contains one feature change: 
  * Delete feature that tried to set the `document.title` with the page detail name. I deleted this feature because it didn't work correctly.


### The most relevant changes are:
* Refactor useUrlStateHooks by batching updates to location
* Delete URL_STATE_KEYS constants. Since inside security threat hitting all pages use the same query string.
* Simplify `x-pack/plugins/security_solution/public/common/components/url_state/index.tsx`
* Transform `dispatchSetInitialStateFromUrl ` into a hook named `useSetInitialStateFromUrl `
* Refactor `use_url_state.tsx`
  * Add `useCallback` to `handleInitialize`
  * Simplify logic in order to improve readability 

### What is NOT included:
* Deleting `updateTimeRange`logic 
* Any change to `<Spyroute>`
* Use router state instead of routeSpy (I don't think that it is possible)
* Use the same router and route spy logic for admin and threat hunting pages
* Move `dispatch(timelineActions.showTimeline` away from `route_capture.tsx`
* Remove `history.replace` from "replace state to location" and use navigate to app instead
* Refactor `makeMapStateToProps` to use hooks as in this commit https://github.com/elastic/kibana/pull/100405/commits/7a4029c628385e13e81e9abd63de66ca5ed06adf


### How to test it?
All possible query strings/filters:
query, filters, savedQuery, sourcerer, timerange, timeline

When we change one of the previously listed filters we rewrite the URL to contain add them. So if we refresh the page the application should persist the data. 
Ex: Open alerts and search for `test`. The URL will update and it should contain `query=(language:kuery,query:test)`

It can be tested in many different ways. And it should be tested for every query string. I am listing here what this code is responsible for:
 * URL updates when changing filter/query/sourcerer/timerange/timeline/savedQuery.
 * When navigating from a page to another the state should be persisted and the new page query string should contain the same query string as the previous page.
 * Remove all query string keys from the URL when navigating to admin. they should be restored when navigating back to threat hunting pages.
 * When initializing the application it should add a query string to the URL with the default filters.
 * When loading a page with the query string in the URL it initializes the APP with preselected values corresponding to the query string.
   

### Checklist

Delete any items that are not applicable to this PR.
